File: src/fsharp/examples/unbound/Scan.fs
(*** hide ***)
module Tutorial.Fs.examples.unbound.Scan

open System
open Microsoft.FSharp.Quotations
open Alea.CUDA
open Alea.CUDA.Utilities 
open Alea.CUDA.Unbound
open NUnit.Framework
open FsUnit

(**
Scan with Alea Unbound. 
*)
[<Test>]
let deviceScanTest() =
    let worker = Worker.Default
    
    let cpuInclusiveScan op (x:'T[]) = Array.scan op x.[0] x.[1..]

    let gpuInclusiveScan (op:Expr<'T -> 'T -> 'T>) (x:'T[]) =
        use scanModule = new DeviceScanModule<'T>(GPUModuleTarget.Worker(worker), op)
        use scan = scanModule.Create(x.Length)
        use dx = worker.Malloc(x)
        use dr = worker.Malloc<'T>(x.Length)
        scan.InclusiveScan(dx.Ptr, dr.Ptr, x.Length)
        dr.Gather()

    let n = 1000
    let rng = Random()
    let x = Array.init n (fun _ -> rng.Next(-5,5))

    let cpuResult = cpuInclusiveScan (+) x
    let gpuResult = gpuInclusiveScan <@(+)@> x
    
    printfn "cpuResult:\n%A" cpuResult
    printfn "gpuResult:\n%A" gpuResult
    gpuResult |> should equal cpuResult

File: src/csharp/examples/unbound/Scan.cs
//[unboundScanImport]
using System;
using System.Collections.Generic;
using System.Linq;
using Alea.CUDA.Unbound;
using NUnit.Framework;
//[/unboundScanImport]

namespace Tutorial.Cs.examples.unbound
{
    public static class Scan
    {
        //[cpuInclusiveScan]
        public static IEnumerable<int> ScanInclusive(this IEnumerable<int> source, int seed, Func<int, int, int> accumulator)
        {
            foreach (var item in source)
            {
                seed = accumulator(seed, item);
                yield return seed;
            }
        }
        //[/cpuInclusiveScan]
    }

    public partial class Test
    {
        //[unboundDeviceScanTest]
        [Test]
        public static void DeviceScanInclusiveTest()
        {
            const int numItems = 1000000;
            var rng = new Random(42);
            var inputs = Enumerable.Range(0, numItems).Select(i => rng.Next(-10, 10)).ToArray();
            var gpuScanModule = DeviceSumScanModuleI32.Default;

            using (var gpuScan = gpuScanModule.Create(numItems))
            using (var dInputs = gpuScanModule.GPUWorker.Malloc(inputs))
            using (var dOutputs = gpuScanModule.GPUWorker.Malloc<int>(inputs.Length))
            {
                gpuScan.InclusiveScan(dInputs.Ptr, dOutputs.Ptr, numItems);
                var actual = dOutputs.Gather();
                var expected = inputs.ScanInclusive(0, (a, b) => a + b).ToArray();
                Assert.AreEqual(expected, actual);
            }
        }
        //[/unboundDeviceScanTest]
    }
}
